http-mclient/src/Network/HTTP/MClient.hs

```diff
@@ -1,156 +1,131 @@
+import qualified Data.ByteString            as BS
 import qualified Data.ByteString.Lazy.Char8 as LBS
-import qualified Data.CaseInsensitive       as CI
 import qualified Data.List                  as List
-import qualified Data.List.NonEmpty         as NE
 import qualified Network.HTTP.Client        as HTTP
-import qualified Network.HTTP.Media         as Media
-
-type Response = HTTP.Response LBS.ByteString
-
-type StatusCode = BoundNumber "Status code" '(199, 600)
+import qualified Network.HTTP.Types         as HTTP
+import qualified UnliftIO.Exception         as Exception
 
 type SendRequest = HTTP.Request -> IO (HTTP.Response LBS.ByteString)
 
 type Env env = HasField "httpSendRequest" env SendRequest
 
-newtype DecodeError = DecodeError Text
-  deriving (Conversion Text) via Text
-  deriving stock (Eq, Show)
-
-data HttpError =
-    ConnectionError HTTP.HttpException
-  | DecodeFailure DecodeError Response
-  | InvalidStatusCode StatusCode Response
-  | InvalidContentType Response
-  | UnsupportedContentType MediaType Response
-  deriving stock (Show)
-
-instance Eq HttpError where
-  DecodeFailure t r          == DecodeFailure t' r'          = t == t' && r == r'
-  InvalidStatusCode code r   == InvalidStatusCode code' r'   = r == r' && code == code'
-  InvalidContentType r       == InvalidContentType r'        = r == r'
-  UnsupportedContentType m r == UnsupportedContentType m' r' = m == m' && r == r'
-  _                          == _                            = False
-
-instance Exception HttpError
-
-data ContentType
-  = FormUrlEncoded
-  | Json
-  | PlainText
-  | XML
-
-class HasMediaType (ctyp :: ContentType) where
-  mediaType :: MediaType
-  mediaType = NE.head (mediaTypes @ctyp)
-
-  mediaTypes :: NE.NonEmpty MediaType
-  mediaTypes = mediaType @ctyp NE.:| []
-
-instance HasMediaType 'FormUrlEncoded where
-  mediaType = "application" // "x-www-form-urlencoded"
-
-instance HasMediaType 'Json where
-  mediaType = "application" // "json" /: ("charset", "utf-8")
-
-instance HasMediaType 'PlainText where
-  mediaTypes
-    = "text" // "plain" /: ("charset", "utf-8")
-    NE.:|
-    [ "text" // "html" /: ("charset", "utf-8")]
-
-mkRequest
-  :: forall contentType acceptsType e a
-  . ( ToText e
-    , HasMediaType contentType
-    , HasMediaType acceptsType
-    )
-  => SendRequest
-  -> (LBS.ByteString -> Either e a)
-  -> HTTP.Request
-  -> IO (Either HttpError a)
-mkRequest sendRequest decodeBody =
-  runExceptT . mkRequest' @contentType @acceptsType (fromType @200) sendRequest decodeBody
-
-mkRequest'
-  :: forall contentType acceptsType a e
-  . ( ToText e
-    , HasMediaType contentType
-    , HasMediaType acceptsType
-    )
-  => StatusCode
-  -> SendRequest
-  -> (LBS.ByteString -> Either e a)
+data ResponseError
+  = BodyDecodeFailure Text
+  | HTTPError HTTP.HttpExceptionContent
+  | UnexpectedContentType BS.ByteString
+  | MissingContentType
+  | UnexpectedStatusCode HTTP.Status
+  deriving stock Show
```

Comment:
Non-blocking: Alternatively, we could have used `deriving anyclass Exception.Exception` for Exception class instance.

Reply:
yeah I agree. Will fix on next touch.